// gojs-angular/esm2020/lib/gojs-angular.module.mjs
import { NgModule } from '@angular/core';
import { DataSyncService } from './data-sync.service';
import { DiagramComponent } from './diagram.component';
import { OverviewComponent } from './overview.component';
import { PaletteComponent } from './palette.component';
import * as i0 from "@angular/core";
export class GojsAngularModule {
}
/** @nocollapse */ GojsAngularModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.1.2", ngImport: i0, type: GojsAngularModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
/** @nocollapse */ GojsAngularModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.1.2", ngImport: i0, type: GojsAngularModule, declarations: [DiagramComponent,
        OverviewComponent,
        PaletteComponent], exports: [DiagramComponent,
        OverviewComponent,
        PaletteComponent] });
/** @nocollapse */ GojsAngularModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.1.2", ngImport: i0, type: GojsAngularModule, providers: [
        DataSyncService
    ] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.1.2", ngImport: i0, type: GojsAngularModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        DiagramComponent,
                        OverviewComponent,
                        PaletteComponent
                    ],
                    imports: [],
                    providers: [
                        DataSyncService
                    ],
                    exports: [
                        DiagramComponent,
                        OverviewComponent,
                        PaletteComponent
                    ]
                }]
        }] });
