// gojs-angular/esm2020/lib/overview.component.mjs
import { Component, ElementRef, Input, NgZone, ViewChild } from '@angular/core';
import * as go from 'gojs';
import { NgDiagramHelper } from "./ng-diagram-helper";
import * as i0 from "@angular/core";
export class OverviewComponent {
    constructor(zone) {
        this.zone = zone;
        /** The Diagram to observe with the Overview */
        this.observedDiagram = null;
        /** The Overview itself  */
        this.overview = null;
    }
    /**
     * Initialize the overview
     */
    ngAfterViewInit() {
        if (!this.overviewDiv) {
            throw new Error("overviewDiv is not defined");
        }
        if (this.initOverview) {
            this.overview = this.initOverview();
            if (!(this.overview instanceof go.Overview)) {
                throw new Error("initOverview function did not return a go.Overview");
            }
        }
        else {
            this.overview = new go.Overview();
            this.overview.contentAlignment = go.Spot.Center;
        }
        // reduces change detection on mouse moves, boosting performance
        NgDiagramHelper.makeMouseMoveRunOutsideAngularZone(this.overview, this.zone);
        this.overview.div = this.overviewDiv.nativeElement;
    }
    /**
     * Only update when the observed diagram changes
     * @param changes
     */
    ngOnChanges(changes) {
        if (!this.overview)
            return;
        if (changes && changes.observedDiagram && changes.observedDiagram.currentValue !== changes.observedDiagram.previousValue) {
            this.overview.observed = changes.observedDiagram.currentValue;
        }
    }
    ngOnDestroy() {
        this.overview.div = null; // removes event listeners
    }
}
/** @nocollapse */ OverviewComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.1.2", ngImport: i0, type: OverviewComponent, deps: [{ token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Component });
/** @nocollapse */ OverviewComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.1.2", type: OverviewComponent, selector: "gojs-overview", inputs: { initOverview: "initOverview", divClassName: "divClassName", observedDiagram: "observedDiagram" }, viewQueries: [{ propertyName: "overviewDiv", first: true, predicate: ["ngOverview"], descendants: true, static: true }], usesOnChanges: true, ngImport: i0, template: '<div #ngOverview [className]=divClassName></div>', isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.1.2", ngImport: i0, type: OverviewComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'gojs-overview',
                    template: '<div #ngOverview [className]=divClassName></div>'
                }]
        }], ctorParameters: function () { return [{ type: i0.NgZone }]; }, propDecorators: { initOverview: [{
                type: Input
            }], divClassName: [{
                type: Input
            }], observedDiagram: [{
                type: Input
            }], overviewDiv: [{
                type: ViewChild,
                args: ['ngOverview', { static: true }]
            }] } });
