// gojs-angular/esm2020/lib/palette.component.mjs
import { Component, ElementRef, EventEmitter, Input, NgZone, Output, ViewChild } from '@angular/core';
import * as go from 'gojs';
import { NgDiagramHelper } from './ng-diagram-helper';
import * as i0 from "@angular/core";
export class PaletteComponent {
    constructor(zone) {
        this.zone = zone;
        /** Link data for palette. Optional. */
        this.linkDataArray = null;
        /** Model data for palette. Optional. */
        this.modelData = null;
        /** Event emitter -- fires when palette model changes. Capture this emitted event in parent component */
        this.modelChange = new EventEmitter();
        /** The Palette itself */
        this.palette = null;
    }
    /**
     * Initialize Palette after view init
     */
    ngAfterViewInit() {
        if (!this.paletteDiv) {
            throw new Error("paletteDiv is not defined");
        }
        this.palette = this.initPalette();
        if (!(this.palette instanceof go.Palette)) {
            throw new Error("initPalette function did not return a go.Palette");
        }
        // reduces change detection on mouse moves, boosting performance
        NgDiagramHelper.makeMouseMoveRunOutsideAngularZone(this.palette, this.zone);
        // assign the Palette's div, which (among many other things) will attach a bunch of listeners to the canvas,
        // using the overridden addEventListener function above
        const divRef = this.paletteDiv.nativeElement;
        if (divRef == null)
            return;
        this.palette.div = divRef;
        // initialize palette model
        NgDiagramHelper.initializeModel(this.palette, this.nodeDataArray, this.linkDataArray, this.modelData);
    }
    /**
     * If a change has occured on an @Input property, merge the app-level changes with GoJS
     */
    ngOnChanges() {
        if (!this.palette || !this.palette.model)
            return;
        NgDiagramHelper.mergeAppDataWithModel(this);
    } // end ngOnChanges 
    ngOnDestroy() {
        this.palette.div = null; // removes event listeners
    }
}
/** @nocollapse */ PaletteComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.1.2", ngImport: i0, type: PaletteComponent, deps: [{ token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Component });
/** @nocollapse */ PaletteComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.1.2", type: PaletteComponent, selector: "gojs-palette", inputs: { initPalette: "initPalette", nodeDataArray: "nodeDataArray", linkDataArray: "linkDataArray", modelData: "modelData", divClassName: "divClassName" }, outputs: { modelChange: "modelChange" }, viewQueries: [{ propertyName: "paletteDiv", first: true, predicate: ["ngPalette"], descendants: true, static: true }], usesOnChanges: true, ngImport: i0, template: '<div #ngPalette [className]=divClassName></div>', isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.1.2", ngImport: i0, type: PaletteComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'gojs-palette',
                    template: '<div #ngPalette [className]=divClassName></div>'
                }]
        }], ctorParameters: function () { return [{ type: i0.NgZone }]; }, propDecorators: { initPalette: [{
                type: Input
            }], nodeDataArray: [{
                type: Input
            }], linkDataArray: [{
                type: Input
            }], modelData: [{
                type: Input
            }], divClassName: [{
                type: Input
            }], modelChange: [{
                type: Output
            }], paletteDiv: [{
                type: ViewChild,
                args: ['ngPalette', { static: true }]
            }] } });
